Rgw admin more actions

maksim-kharitonov opened the pull request:
<!--
Thank you for opening a pull request. Please provide:

- A clear summary of your changes

- Descriptive and succinct commit messages with the format:
  """
  [topic]: [short description]

  [Longer description]

  Signed-off-by: [Your Name] <[your email]>
  """

  Topic will generally be the go ceph package dir you are working in.

- Ensure checklist items listed below are accounted for
-->

This change implements 
- listing of user's buckets (also with stats), 
- user credentials management, 
- individual bucket quota APIs 
of rgw/admin.

Related to https://github.com/ceph/go-ceph/issues/653

Signed-off-by: maxharitonov [maxharitonov@gmail.com](mailto:maxharitonov@gmail.com)

## Checklist
- [x] Added tests for features and functional changes
- [x] Public functions and types are documented
- [x] Standard formatting is applied to Go code
- [x] Is this a new API? Is this new API marked PREVIEW?

phlogistonjohn commented on rgw/admin/bucket.go:
```diff
@@ -96,6 +96,66 @@ func (api *API) ListBuckets(ctx context.Context) ([]string, error) {
+}
+
+// ListUsersBuckets will return the list of all users buckets without stat
+//  PREVIEW
```
Sorry if the docs don't currently make this clear but new functions added as "PREVIEW" must also be "protected" by a build tag "ceph_preview". Unfortunately, this means you can't currently add a new API function to an existing file with stable functions in it.

Also tagging @leseb for attention for these rgw api changes.

maksim-kharitonov replied:
@phlogistonjohn thank you for the feedback. I've added commit, moving new API methods to separate files with `+build ceph_preview`

leseb commented on rgw/admin/user_bucket_test.go:
```diff
@@ -0,0 +1,58 @@
+		assert.Equal(suite.T(), 1, len(buckets))
+	})
+
+	suite.T().Run("list unknown user's buckets", func(t *testing.T) {
```
```suggestion
	suite.T().Run("list unknown user's buckets with stat", func(t *testing.T) {
```

leseb commented on rgw/admin/user_bucket.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+// ListUsersBucketsWithStat will return the list of all users buckets with stat
+//  PREVIEW
```
```suggestion
// PREVIEW
```

leseb commented on rgw/admin/user_bucket.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+// ListUsersBuckets will return the list of all users buckets without stat
+//  PREVIEW
```
```suggestion
// PREVIEW
```

leseb commented on rgw/admin/keys.go:
```diff
@@ -0,0 +1,64 @@
+// RemoveKey will remove an existing key
+// https://docs.ceph.com/en/latest/radosgw/adminops/#remove-key
+// KeySpec.SecretKey parameter shouldn't be provided and will be ignored
+//  PREVIEW
```
```suggestion
// PREVIEW
```

leseb commented on rgw/admin/keys.go:
```diff
@@ -0,0 +1,64 @@
+
+// CreateKey will generate new keys or add specified to keyring
+// https://docs.ceph.com/en/latest/radosgw/adminops/#create-key
+//  PREVIEW
```
```suggestion
// PREVIEW
```

leseb commented on rgw/admin/bucket_quota_test.go:
```diff
@@ -0,0 +1,27 @@
+		err := co.SetIndividualBucketQuota(context.Background(), IndividualBucketQuotaSpec{UID: "admin"})
+		assert.Error(suite.T(), err)
+		assert.EqualError(suite.T(), err, errMissingUserBucket.Error())
+	})
```
Where is the successful test?

maksim-kharitonov replied:
@leseb thank you for your review and fixes. Added commit with positive case test and fix of idempotentness of keys API tests

leseb commented on docs/api-status.md:
```diff
@@ -56,6 +56,11 @@ Snapshot.Set | v0.10.0 |  |
 
 Name | Added in Version | Expected Stable Version | 
 ---- | ---------------- | ----------------------- | 
+API.ListUsersBuckets | v0.13.0 | v0.14.0 | 
```
Should this be Added in 0.14 and Expected stable in 0.15? @phlogistonjohn

phlogistonjohn replied:
Thanks for noticing that. We're aware of the workflow issues with the preview policy and "stale" versions numbers in docs/json.
The "Added in Version" should always be the upcoming go-ceph release, so in this case "v0.15.0". The script automatically puts Y+2 for stable, so "v0.17.0" but that is negotiable.

maksim-kharitonov replied:
fixed

leseb commented on rgw/admin/bucket.go:
```diff
@@ -112,7 +112,7 @@ func (api *API) GetBucketInfo(ctx context.Context, bucket Bucket) (Bucket, error
 }
 
-// GetBucketPolicy - http://docs.ceph.com/docs/mimic/radosgw/adminops/#get-bucket-or-object-policy
+// GetBucketPolicy - http://docs.ceph.com/en/mimic/radosgw/adminops/#get-bucket-or-object-policy
```
Since we are changing this line, should we link `latest` instead of `mimic` too?

maksim-kharitonov replied:
fixed

leseb commented on rgw/admin/bucket_quota_test.go:
```diff
@@ -0,0 +1,38 @@
+
+	suite.T().Run("set bucket quota", func(t *testing.T) {
+		err := co.SetIndividualBucketQuota(context.Background(), IndividualBucketQuotaSpec{UID: "admin", Bucket: suite.bucketTestName})
+		assert.NoError(suite.T(), err)
```
What can we do despite checking for the `SetIndividualBucketQuota` to succeed to verify that the individual quota has been set properly? Is there a "get" we can use and read from?

leseb commented on rgw/admin/bucket_quota.go:
```diff
@@ -0,0 +1,44 @@
+
+// SetIndividualBucketQuota sets quota to a specific bucket
+// https://docs.ceph.com/en/latest/radosgw/adminops/#set-quota-for-an-individual-bucket
+//  PREVIEW
```
```suggestion
// PREVIEW
```

leseb commented on rgw/admin/keys.go:
```diff
@@ -0,0 +1,64 @@
+// https://docs.ceph.com/en/latest/radosgw/adminops/#create-key
+// PREVIEW
+func (api *API) CreateKey(ctx context.Context, key KeySpec) (*[]UserKeySpec, error) {
+	if key.UID == "" && key.SubUser == "" {
```
Should UID and SubUser be treated with different errors instead?

maksim-kharitonov replied:
I suppose it's just the same error - if neither UID neither nor SubUser provided it's an error, if any exists it's ok

leseb commented on rgw/admin/keys.go:
```diff
@@ -0,0 +1,64 @@
+// KeySpec.SecretKey parameter shouldn't be provided and will be ignored
+// PREVIEW
+func (api *API) RemoveKey(ctx context.Context, key KeySpec) error {
+	if key.UID == "" && key.SubUser == "" {
```
ditto

leseb reviewed (commented):
Just minor comments, almost there, thanks!

sebastianriese commented:
Sorry for creating conflicts with your PR by having my changes merged. I hope I can be helpful by giving hints on what must be done to get this code running again when rebasing it atop master which now includes #644.

The `valueToURLParams` function now requires a second parameter of type `[]string` that lists the allowed parameters for the API calls. The allowed values are not fully documented in the API docs, so the best way to extract the full list for each API call is to consult the ceph-rgw source code. The relevant files are the ones called `rgw_rest_*.cc` in the path <https://github.com/ceph/ceph/tree/master/src/rgw>. E.g. the following block parses the parameters for the bucket link operation: <https://github.com/ceph/ceph/blob/master/src/rgw/rgw_rest_bucket.cc#L142> – and you can read off which parameters are interpreted by the ceph rgw admin ops API.

The convention now is not to include parameters in the allowed parameter list, that we don't currently support in our data structures, but to document this in a comment like here: <https://github.com/ceph/go-ceph/pull/644/files#diff-694c344990ad130009eec2ff3c457832c960d480dfee9d1b675a2ea6138c5990R120>.

phlogistonjohn commented:
Hello there, if you wish us to revisit this PR please manually resolve the conflicts. Also, a new version has been released so the versions in docs will need updating. We're looking to simplify this process but at the moment it still requires manual fixing.

maksim-kharitonov commented:
> Hello there, if you wish us to revisit this PR please manually resolve the conflicts. Also, a new version has been released so the versions in docs will need updating. We're looking to simplify this process but at the moment it still requires manual fixing.

Thank you for your assistance. Sorry for the delay in fixing, I'll try to come back with edits asap

phlogistonjohn commented:
Hi @maksim-kharitonov  - Just doing a ping in case you are still planning on working on this.  I'd suggest that if you don't think you'll get to it in the next month or so, we can close this PR - and you can always reopen it in the future when you are ready.
Having it closed helps us maintainers not have to do things like pester you with messages like these and it keeps our list of things to look at smaller. 
Thanks!

maksim-kharitonov commented:
> Hi @maksim-kharitonov - Just doing a ping in case you are still planning on working on this. I'd suggest that if you don't think you'll get to it in the next month or so, we can close this PR - and you can always reopen it in the future when you are ready. Having it closed helps us maintainers not have to do things like pester you with messages like these and it keeps our list of things to look at smaller. Thanks!

Thank you for your patience, just resolved conflicts and fixed. Can you please re-check pr?

thotz commented on rgw/admin/user_bucket.go:
```diff
@@ -0,0 +1,71 @@
+	}
+
+	generateStat := false
+	listingSpec := BucketListingSpec{
```
IMO `listingSpec` is not required for this API, just pass the uid directly to request

maksim-kharitonov replied:
and what about `stats` query parameter?

thotz replied:
IMO that parameter is not required

maksim-kharitonov commented on rgw/admin/bucket_quota.go:
```diff
@@ -0,0 +1,44 @@
+//go:build ceph_preview
+// +build ceph_preview
+
+package admin
```
according to https://github.com/ceph/go-ceph/pull/654#discussion_r811275166 I can't currently add a new API function to an existing file with stable functions in it.

thotz replied:
@phlogistonjohn @ansiwen preview tag is for the API right, is it per file? That means we cannot add new apis to existing files

@maksim-kharitonov : Thanks for explaining the API issue, but still IMO u can extend the existing struct than defining new ones for your APIs

maksim-kharitonov replied:
@thotz ok, I've reused `QuotaSpec` type for individual bucket quota

phlogistonjohn replied:
That's correct - build tags are per-file. Do note that Go let's you implement receivers (methods) for types across many files.  So you can extend existing types with preview apis by putting those methods in new files.

maksim-kharitonov commented on rgw/admin/keys.go:
```diff
@@ -0,0 +1,64 @@
+
+// KeySpec is the user credential management configuration
+// https://docs.ceph.com/en/latest/radosgw/adminops/#id37
+type KeySpec struct {
```
`UserSpecKey` has a field with `json:"user"`  https://github.com/ceph/go-ceph/blob/master/rgw/admin/user.go#L91 is it a good idea to add to the field annotation `url:"uid"` ?

thotz replied:
Same scenario can be seen with `user` as well https://github.com/ceph/go-ceph/blob/master/rgw/admin/user.go#L12
IMO field name in struct, URL request entry and json response entry all can have different value

maksim-kharitonov replied:
Not exactly the same: `KeySpec` also has `SubUser` field. Key create response can be parsed to `UserKeySpec`, but I don't want to add subuser to `UserKeySpec`. Another option is to add two methods: CreateKey: UserKeySpec->UserKeySpec and CreateSubUserKey: SwiftKeySpec->SwiftKeySpec but it also doesn't look good.

thotz replied:
Then define `user` separately, not use with `UID`. Currently UserKeySpec holds only response, just add request parameters as well
```go
type UserKeySpec struct {
	User      string `json:"user"`
	AccessKey string `json:"access_key" url:"access-key"`
	SecretKey string `json:"secret_key" url:"secret-key"`
        UID         string `url:"uid"`          // The user ID to receive the new key
	SubUser     string `url:"subuser"`      // The subuser ID to receive the new key
	KeyType     string `url:"key-type"`
        GenerateKey *bool  `url:"generate-key"` // Generate a new key pair and add to the existing keyring
}
````
It should be for both cases IMO

maksim-kharitonov replied:
fixed

maksim-kharitonov commented on rgw/admin/bucket_quota.go:
```diff
@@ -0,0 +1,44 @@
+//  PREVIEW
+func (api *API) SetIndividualBucketQuota(ctx context.Context, quota IndividualBucketQuotaSpec) error {
+	// Always for quota type to bucket
+	quota.QuotaType = "bucket"
```
fixed

maksim-kharitonov commented on rgw/admin/keys.go:
```diff
@@ -0,0 +1,64 @@
+	if key.UID == "" && key.SubUser == "" {
+		return nil, errMissingUserID
+	}
+
```
added detailed analysis of KeyType and parameters

maksim-kharitonov commented on rgw/admin/keys.go:
```diff
@@ -0,0 +1,64 @@
+	if key.UID != "" && key.AccessKey == "" {
+		return errMissingUserAccessKey
+	}
+
```
the same

thotz commented on rgw/admin/bucket_quota.go:
```diff
@@ -0,0 +1,44 @@
+	}
+
+	return nil
+}
```
If possible I was refering to add new API for setting bucket qota on a user

maksim-kharitonov replied:
I think it already exists in `SetUserQuota` method, it has `QuotaSpec.QuotaType` parameter

phlogistonjohn commented:
FYI, things have changed a little around preview api documentation since you first filed the PR. The "PREVIEW" term in the comment is no longer needed. Instead, preview status is entirely derived from the build tag. Please remove the "PREVIEW" lines from the comments when you have a moment.

thotz commented on rgw/admin/user_bucket_test.go:
```diff
@@ -0,0 +1,58 @@
+		assert.True(suite.T(), errors.Is(err, errMissingUserID), err)
+	})
+
+	suite.T().Run("list user's buckets with stat", func(t *testing.T) {
```
we may need to validate stats here to have a complete check of API. And some entires such as pool is missing from current bucket struct in go-ceph. May not needed for API, but useful for future I guess

maksim-kharitonov replied:
added a few checks on user's buckets list stats

thotz reviewed (commented):
Almost there need to update the test case for listbucketstat, apart from that PR LGTM

anoopcs9 commented:
@maksim-kharitonov

It looks better if you can squash various commits in situations like:
* Addressing review comments
* Fixing style issues
* Adding new APIs via individual files (just one or two commits adding particular set of APIs and tests)

Let me know if you need more clarification.

maksim-kharitonov commented:
> @maksim-kharitonov
> 
> It looks better if you can squash various commits in situations like:
> 
> * Addressing review comments
> * Fixing style issues
> * Adding new APIs via individual files (just one or two commits adding particular set of APIs and tests)
> 
> Let me know if you need more clarification.

@anoopcs9 rebased my pr to new master and squashed to single commit

anoopcs9 commented:
> > @maksim-kharitonov
> > It looks better if you can squash various commits in situations like:
> > 
> > * Addressing review comments
> > * Fixing style issues
> > * Adding new APIs via individual files (just one or two commits adding particular set of APIs and tests)
> > 
> > Let me know if you need more clarification.
> 
> @anoopcs9 rebased my pr to new master and squashed to single commit

I'm sorry if I was not clear earlier. I didn't mean to squash everything into one large commit rather have meaningful and related changes included as fewer individual commits like:
* _bucket_quota_ API addition and tests(with related changes to existing sources)
* _user_bucket_ API addition and tests(with related changes to existing sources)
* _keys_ API addition and tests(with related changes to existing sources)
* docs change for `api-status`

Also please note that `ceph_preview` build tags are still missing from newly added _bucket_quota_test.go_, _keys_test.go_ and _user_bucket_test.go_ sources.

And finally changes to _bucket.go_ have rejected hunks when applied on current master. Please rebase and resolve correctly on latest master.

maksim-kharitonov commented:
> > > @maksim-kharitonov
> > > It looks better if you can squash various commits in situations like:
> > > 
> > > * Addressing review comments
> > > * Fixing style issues
> > > * Adding new APIs via individual files (just one or two commits adding particular set of APIs and tests)
> > > 
> > > Let me know if you need more clarification.
> > 
> > 
> > @anoopcs9 rebased my pr to new master and squashed to single commit
> 
> I'm sorry if I was not clear earlier. I didn't mean to squash everything into one large commit rather have meaningful and related changes included as fewer individual commits like:
> 
> * _bucket_quota_ API addition and tests(with related changes to existing sources)
> * _user_bucket_ API addition and tests(with related changes to existing sources)
> * _keys_ API addition and tests(with related changes to existing sources)
> * docs change for `api-status`
> 
> Also please note that `ceph_preview` build tags are still missing from newly added _bucket_quota_test.go_, _keys_test.go_ and _user_bucket_test.go_ sources.
> 
> And finally changes to _bucket.go_ have rejected hunks when applied on current master. Please rebase and resolve correctly on latest master.

1. splitted into commits with more explicit meaning
2. added tags to test-files
3. rebased to the latest master

anoopcs9 commented on docs/api-status.md:
```diff
@@ -79,4 +79,8 @@ Admin.CreateCephFSExport | v0.16.0 | v0.18.0 |
 Admin.ListDetailedExports | v0.16.0 | v0.18.0 | 
 Admin.ExportInfo | v0.16.0 | v0.18.0 | 
-
+API.ListUsersBuckets | v0.16.0 | v0.18.0 | 
+API.ListUsersBucketsWithStat | v0.16.0 | v0.18.0 | 
+API.CreateKey | v0.16.0 | v0.18.0 | 
+API.RemoveKey | v0.16.0 | v0.18.0 | 
+API.SetIndividualBucketQuota | v0.16.0 | v0.18.0 | 
```
New preview APIs are added under wrong(**Package: common/admin/nfs**) section. Running `make api-update` should ensure that these are placed correctly under corresponding package headers.

maksim-kharitonov replied:
fixed

anoopcs9 reviewed (changes requested):
Almost there...please see my comment.

anoopcs9 commented:
Just noticed that commit messages are not proper(not visible in a single line). Can we have that also corrected please? Sorry for not pointing it out earlier.
```
rgw/admin: Individual bucket quota support
rgw/admin: Key create and delete api support
rgw/admin: List user's bucket support
docs: Update API status

maksim-kharitonov commented:
> Just noticed that commit messages are not proper(not visible in a single line). Can we have that also corrected please? Sorry for not pointing it out earlier.
> 
> ```
> rgw/admin: Individual bucket quota support
> rgw/admin: Key create and delete api support
> rgw/admin: List user's bucket support
> docs: Update API status
> ```

fixed

phlogistonjohn commented on docs/api-status.json:
```diff
@@ -1761,6 +1761,36 @@
+      {
+        "name": "API.ListUsersBuckets",
+        "comment": "ListUsersBuckets will return the list of all users buckets without stat.\n",
+        "added_in_version": "v0.16.0",
```
Sorry but this needs to be fixed, the v0.16 release has already occurred so this would need to be updated to v0.17.0.  I am aware that this is a pain and I plan on tweaking the process so contributors do not need to do this in the future. Unfortunately, for now you need to fix it.

maksim-kharitonov replied:
fixed version to v0.17.0

anoopcs9 replied:
> Sorry but this needs to be fixed, the v0.16 release has already occurred so this would need to be updated to v0.17.0. I am aware that this is a pain and I plan on tweaking the process so contributors do not need to do this in the future. Unfortunately, for now you need to fix it.

In that case **expected_stable_version** would be `v0.19.0`, right?

maksim-kharitonov replied:
yes, fixed version

phlogistonjohn reviewed (changes requested):
The API json (and doc) are not up to date. You can either manually fix both files or you can try resetting the patch and re-running `make api-update`.

phlogistonjohn commented:
Please note that the CI is experiencing issues do to a problem in the most recent ceph quincy hotfix release. If it's not fixed by ceph upstream soon, I plan to apply a workaround. Thanks for your patience.

anoopcs9 commented:
> Please note that the CI is experiencing issues do to a problem in the most recent ceph quincy hotfix release. If it's not fixed by ceph upstream soon, I plan to apply a workaround. Thanks for your patience.

This is now fixed with latest Quincy release v17.2.3 

@maksim-kharitonov If possible can you also verify outstanding unresolved conversations and take action accordingly(mark as resolved)?

phlogistonjohn reviewed (approved):
LGTM, thanks for following up on the annoying managerial tasks.

phlogistonjohn commented:
@mergifyio rebase
